### DISCUSSION TITLE
Writing a second, cross-platform GUI for vAmiga?

### DISCUSSION THREAD

**Discussion by dirkwhoffmann** [dirkwhoffmann/vAmiga#525] [Ideas]
During the last couple of days, I was thinking about the pros and cons of writing a second GUI for vAmiga. The major goal would be to make vAmiga accessible on Windows and Linux. 

A couple of month ago, I experimented with SDL and SFML which resulted in a proof-of-concept emulator called vAmigaBM (vAmiga Bare Metal). However, I feel that this is a dead-end. The major selling point of vAmiga is it’s sophisticated GUI (I think) and this selling point is just no longer there in vAmigaBM.

I browsed the web for a quick overview of cross-platform GUI frameworks. QT and GTK seem to be the two big players, but I also noticed that wxWidgets got some positive reviews.

My current understanding is:

- QT is massive. Furthermore, it only mimics the native look and feel by using skins.
- GTK is Linux centric and will definitely look like an app from outer space when run on Windows.
- wxWidgets is lightweight and maps to the real GUI elements of the OS.

From this quick analysis, wxWidgets seems like a suitable candidate, since I personally dislike apps that don’t exhibit a native look and feel. However, there are other requirements: The frameworks needs to be fast (at least for keyboard and mouse events), and it needs to be capable of handling complex UI interfaces such as the Inspector window with all of it’s sub panels.

Every feedback is welcome, especially if you are aware of any reasons why choosing wxWidgets would be the wrong decision.

**Comment by shinra-electric** [comment]
I have no experience with any of these, but I know that PCSX2 is planning to remove wxWidgets. I can't seem to find any post that detail their reasonings, but you can see that it's planned from pull requests like [this](https://github.com/PCSX2/pcsx2/pull/4707). You might want to go to the dev channel on their discord and ask why. 

I hope this remains a "second" GUI, and does not become the primary. The native Mac interface is the main reason why I love vAmiga and VirtualC64... 


Speaking of the Mac GUI, I've been making a SwiftUI interface for vAmiga just for fun. I kind of stalled when I got to the table views, but now that tables are supported in SwiftUI 3 I might get stuck back in and try to finish it off. 

Here's what I've done so far, if you're interested...

[vAmigaUI.zip](https://github.com/dirkwhoffmann/vAmiga/files/7147522/vAmigaUI.zip)

**Comment by dirkwhoffmann** [comment]
> I hope this remains a "second" GUI, and does not become the primary

Yes, if I ever write it, it will be a second GUI, at least as long as I stay on the Mac platform. If Apple does continue on the path it has taken (e.g., locking down macOS more and more,  implementing their plans to actively scan our devices, etc.) I will say good bye at some point. 

I just run the SwiftUI version. Interestingly, it doesn't look like a real native macOS app. For some reason, many controls look different. SwiftUI is another Apple move I didn't really get. But maybe the technology is just still premature yet.

**Comment by shinra-electric** [comment]
> Yes, if I ever write it, it will be a second GUI, at least as long as I stay on the Mac platform. If Apple does continue on the path it has taken (e.g., locking down macOS more and more, implementing their plans to actively scan our devices, etc.) I will say good bye at some point.

Well, I'd agree in principle, though I don't think that's what they are really doing. Federighi has said many times it's going to remain open. Also, terrible reporting has mis-represented what that CSAM thing actually does. And it looks like they are backing down a bit now, but in any case this is off topic...
 
> I just run the SwiftUI version. Interestingly, it doesn't look like a real native macOS app. For some reason, many controls look different. SwiftUI is another Apple move I didn't really get. But maybe the technology is just still premature yet.

Some of the controls do look a bit different, but this may be down to me messing around, trying different things. I couldn't adjust  the size of the slider control, but I believe it's possible now. And I tried to use SF Symbols or (emoji) rather than images. 

They are definitely focusing on iOS over macOS right now. There are still lots of things that are unimplemented or are just terrible to use right now. But it's going to improve, and has already improved a huge amount since the first release. 

SwiftUI in general is great though. Not only is it easy to use, but it's cross-(Apple)platform (with a little bit of work). 

But again, this is off-topic. We should be talking about the cross-platform stuff....

**Comment by prb28** [comment]
Hi Dirk,
The main users of the emulators, as games are on windows as you noted.
I like your idea to bring the the emulator windows and linux, and the graphical interface is a major point for most of the users. I don't know what makes a user choose an emulator, certainly simplicity and accuracy, but I might be wrong.
I've tried to build vAmigaBM on windows msvc and it was not as easy as expected.
Did you try it ?

For the GUI, I think you do not need something really complex, so the lightest and simplest to build and use should be the best. I dealt with cross platform build sometimes and it is always a headache.
I think GTK may be more updated (supported) than WxWindows. Qt seems to me too big for the purpose (the project has a lot of dependencies and need some custom install to be ready to build).
Maybe SFML is enough.

In addition to the selection of the library used to manage the GUI you should consider an easy way to build to project in all the plate-forms with some kind of continuous integration.
Do you have some ideas in how you want to handle this ?

**Reply by mithrendal** [reply]
Offtopic: bon soir @prb28 , ce vachement cool que tu as repondu ici ! 😎👍🏻 I first thought github has a problem and mixes up all the different projects 😅 as I also follow your very impressive  https://github.com/prb28/vscode-amiga-assembly

**Reply by prb28** [reply]
@mithrendal my German is so bad that I would not even try to answer with a phrase. ;-)
Yes I follow this project a little far away since I've replace my mac as hobby computer by a dirty windows... I know...
But somewhere in my mind, the simplicity of this emulator seems great, I was looking if I can build it (just trying now and getting frustrated with the enum definition) maybe contributing with a remote gdb protocol if @dirkwhoffmann is interested.

**Reply by dirkwhoffmann** [reply]
Oh, we have the author of the famous amiga assembly plugin as a guest in our forum. That just completely escaped my attention. Welcome on board!

**Reply by prb28** [reply]
Famous... that's too much...
I've seen that you have @emmon on board too. I've started my debugging project on his work with fs-uae, so he may already have an idea of integrating it with some of his tools.

**Reply by dirkwhoffmann** [reply]
> remote gdb protocol if @dirkwhoffmann is interested

@prb28 Yes, I am definitely interested. I think integrating vAmiga in VS would be super exciting. Right now, I have no idea what changes in vAmiga will be necessary, but I am eager to work on this project.

**Reply by prb28** [reply]
@dirkwhoffmann It's easy for the protocol part, the tricky part is the integration of the commands with your emulation.
After your build on windows we can work together on it.
To have an idea of the integration in WinUAE (done by bartman and modified by me): https://github.com/prb28/WinUAE/blob/master/od-win32/barto_gdbserver.cpp
Or in FS-UAE (done by emoon and modified by me): https://github.com/prb28/fs-uae/tree/master/src/remote_debug

**Reply by dirkwhoffmann** [reply]
OK, great! Let's work on this once I've finished the Windows compatibility project.

**Comment by dirkwhoffmann** [comment]
Hi @prb28,

vAmigaBM is basically dead, because SFML has turned out to be a dead end. It's not like SFML was a bad framework. It delivered what it promised, but I've learned that GUI elements such as menus, buttons, tables, and all kinds of other widgets are essential to provide a good emulator experience. In addition, vAmigaBM uses an outdated emulator core, and I don't think it's worth to make it work with the recent one. I am even thinking about deleting the repo. 

Regarding Windows builds, you might be interested in this thread: https://github.com/dirkwhoffmann/vAmiga/issues/611

Making the code base compatible with Windows is what I currently do. If everything works out as expected, it should be possible to build the core emulator without major issues in a couple of days (please note that there is no Windows GUI at the moment. It's all about compiling the core emulator in MSVC).

**Reply by prb28** [reply]
Very interesting, I'll follow this issue to see your progress.
(I'll try to build it with your branch)
